examples/fluids/src/qdata.c

```diff
@@ -0,0 +1,199 @@
+// Copyright (c) 2017-2024, Lawrence Livermore National Security, LLC and other CEED contributors.
+// All Rights Reserved. See the top-level LICENSE and NOTICE files for details.
+//
+// SPDX-License-Identifier: BSD-2-Clause
+//
+// This file is part of CEED:  http://github.com/ceed
+
+#include "../navierstokes.h"
+
+#include <petscsection.h>
+#include "../qfunctions/setupgeo.h"
+#include "../qfunctions/setupgeo2d.h"
+
+/**
+ * @brief Get number of components of quadrature data for domain
+ *
+ * @param[in]  dm          DM where quadrature data would be used
+ * @param[out] q_data_size Number of components of quadrature data
+ */
+PetscErrorCode QDataGetNumComponents(DM dm, CeedInt *q_data_size) {
+  PetscInt num_comp_x, dim;
+
+  PetscFunctionBeginUser;
+  PetscCall(DMGetDimension(dm, &dim));
+  {  // Get number of coordinate components
+    DM           dm_coord;
+    PetscSection section_coord;
+    PetscInt     field = 0;  // Default field has the coordinates
+    PetscCall(DMGetCoordinateDM(dm, &dm_coord));
+    PetscCall(DMGetLocalSection(dm_coord, &section_coord));
+    PetscCall(PetscSectionGetFieldComponents(section_coord, field, &num_comp_x));
+  }
+  switch (dim) {
+    case 2:
+      switch (num_comp_x) {
+        case 2:
+          *q_data_size = 5;
+          break;
+        case 3:
+          *q_data_size = 7;
+          break;
+        default:
+          SETERRQ(PetscObjectComm((PetscObject)dm), PETSC_ERR_SUP,
+                  "QData not valid for DM of dimension %" PetscInt_FMT " and coordinates with dimension %" PetscInt_FMT, dim, num_comp_x);
+          break;
+      }
+      break;
+    case 3:
+      *q_data_size = 10;
+      break;
+    default:
+      SETERRQ(PetscObjectComm((PetscObject)dm), PETSC_ERR_SUP,
+              "QData not valid for DM of dimension %" PetscInt_FMT " and coordinates with dimension %" PetscInt_FMT, dim, num_comp_x);
+      break;
+  }
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+/**
+ * @brief Create quadrature data for domain
+ *
+ * @param[in]  ceed          Ceed object quadrature data will be used with
+ * @param[in]  dm            DM where quadrature data would be used
+ * @param[in]  domain_label  DMLabel that quadrature data would be used one
+ * @param[in]  label_value   Value of label
+ * @param[in]  elem_restr_x  CeedElemRestriction of the coordinates (must match `domain_label` and `label_value` selections)
+ * @param[in]  basis_x       CeedBasis of the coordinates
+ * @param[in]  x_coord       CeedVector of the coordinates
+ * @param[out] elem_restr_qd CeedElemRestriction of the quadrature data
+ * @param[out] q_data        CeedVector of the quadrature data
+ * @param[out] q_data_size   number of components of quadrature data
+ */
+PetscErrorCode QDataGet(Ceed ceed, DM dm, DMLabel domain_label, PetscInt label_value, CeedElemRestriction elem_restr_x, CeedBasis basis_x,
```

Comment:
Is there ever a point in which you will call `QDataGet` for the same volumetric region multiple times and might  want to re-use the q-data?

Reply:
Yep. That would help remove `ceed_data->elem_restr_qd_i` and `ceed_data->q_data`. I originally had this named `QDataCreate`, but realized I'd probably want to do that in the future, so changed to name to `Get`.

Primary roadblock there is how to ID a `{elem_restr_qdata, q_data, q_data_size}` tuple. I don't want to ID off of the DM, as many DMClones can (and currently do) use the same qdata. The other thought would be to always create a `elem_restr_qd`, then ID off of that, but idk if we have a `CeedElemRestrictionCompare` or a way to get equivalent functionality.

Reply:
Or, always create the `q_data` vector, then compare values (with some tolerance). A bit expensive, but pretty much fool proof.

Reply:
We have an object corresponding to material region in Ratel and that object keeps track of the QData but I'm not entirely satisfied with it as a general purpose approach. I do think ultimately we'll want to attach to the DM as part of the eventual upstreaming to PETSc. Its not a blocker here but definitely something to think about as we are maturing the integration between PETSc and libCEED

Reply:
For Ratel the material region keeps track of its own boundary since some of our boundary operators rely upon the material model selected for the region